glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,357 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Ping upload helper.
+
+// !IMPORTANT!
+// Remove the next line when this module's
+// functionality is in the Glean object. This
+// is here just to not have lint error for now.
+#![allow(dead_code)]
+
+use std::fs::{remove_dir_all, remove_file, File};
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+
+use lazy_static::lazy_static;
+use log;
+use regex::Regex;
+use serde_json::Value as JsonValue;
+use uuid::Uuid;
+
+use crate::Result;
+use queue::PingsQueue;
+use request::PingRequest;
+
+mod queue;
+mod request;
+
+lazy_static! {
+    /// Ping file names are UUIDs, this matches UUIDs for filtering purposes
+    static ref PING_FILE_NAME_PATTERN: Regex = Regex::new("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}").unwrap();
+}
+
+/// Attempt to delete a file.
+///
+/// This won't panic if not able to delete, just log.
+fn delete_file(path: &Path) {
+    match remove_file(path) {
+        Err(e) => log::error!("Error deleting file {}. {}", path.display(), e),
+        _ => log::info!("Deleted file {}", path.display()),
+    };
+}
+
+/// Get the file name from a path as a &str.
+///
+/// This won't panic if not able to get file name, just log.
+fn get_file_name_as_str(path: &Path) -> Option<&str> {
+    match path.file_name() {
+        None => {
+            log::warn!("Error getting file name from path: {}", path.display());
+            None
+        }
+        Some(file_name) => {
+            let file_name = file_name.to_str();
+            if file_name.is_none() {
+                log::warn!("File name is not valid unicode: {}", path.display());
+            }
+            file_name
+        }
+    }
+}
+
+/// Creates requests for ping upload and process responses from them.
+pub struct PingUploadHelper {
+    /// A queue of ping UUIDs to upload
+    queue: PingsQueue,
+    /// The path to the pending_pings directory
+    pending_pings_dir: PathBuf,
+}
+
+impl PingUploadHelper {
+    /// Create a new PingUploadHelper.
+    ///
+    /// Starts an empty queue to be filled with the pending pings
+    /// file paths that should be processed and uploaded next.
+    pub fn new(data_path: &Path) -> Self {
+        Self {
+            queue: PingsQueue::new(),
+            pending_pings_dir: data_path.join("pending_pings"),
+        }
+    }
+
+    /// Creates a `PingRequest` for the first possible ping in the queue.
+    ///
+    /// If one of the files is malformed or generates errors, it will
+    /// keep trying while there are still pending pings.
+    pub fn create_next_ping_request(&self) -> Option<PingRequest> {
```

Comment:
Is this the function that needs to be called from the outside to get any available work?

If so, please note that, [by design](https://docs.google.com/document/d/17D5K--z8GC-vU8t1sKwziTwexQTtDzYB99xDzhUIjiI/edit#heading=h.ap1pjdulp5vm), this should return whether or not work is available or if the caller needs to wait. Are you planning on doing that in a separate bug?

Reply:
Yes, I am. I thought about this, but for the stage the module is in right now, I can't see a situation where we would tell the caller to wait. Please correct me if I am wrong.

Upon implementing, I actually realized that I am not sure at which time we would tell the caller to wait other than for throttling purposes, so I thought we could do this when looking into that.

Reply:
That's fine, I wanted to make sure about that :)

Reply:
Why do you think so?

Reply:
Because naming is hard :-)

What is this API doing in the first place? We're calling it to understand if there's anything to upload or if we need to wait. I'm not saying `get_ping_to_upload` is a good name, but I'm not sure `create_next_ping_request` is as well!

The `next` within the function name reminds me of some Win32 APIs that were like "GetFirstSomething" / "GetNextSomething": the next implies, for me, that there was a "first" :) In addition to that, the fact that it creates a "request" object is an implementation detail: the real objective of calling this function is, really, just getting the indications on what to do (upload or wait)

Reply:
> Because naming is hard :-)

LOL

---

This is very interesting to me, we understand the name so differently! I agree with your argument about the request part being an implementation detail and could be left out of the function name. But the `next` to me in necessary because as I understand it, it implies that there is more than one and we should keep coming back until there is none.

It's like "Next, please!" when you are calling the next person in a line.

My suggestion is to rename it to `get_next_ping_to_upload`, merging our two suggestions.

Reply:
> My suggestion is to rename it to `get_next_ping_to_upload`, merging our two suggestions.

Sounds good to me :-) Let's see what @badboy thinks :)